src/java/org/apache/cassandra/index/sai/plan/QueryController.java

```diff
@@ -206,136 +219,135 @@ public UnfilteredRowIterator getPartition(PrimaryKey key, ReadExecutionControlle
-
-            queryContext.sstablesHit += queryView.referencedIndexes
-                                        .stream()
-                                        .map(SSTableIndex::getSSTable).collect(Collectors.toSet()).size();
-            queryContext.checkpoint();
-            RangeIterator<PrimaryKey> union = RangeUnionIterator.build(allIntersections);
-            return new CheckpointingIterator<>(union,
-                                               queryView.referencedIndexes,
-                                               annQueryViewInHybridSearch == null ? Collections.emptySet() : annQueryViewInHybridSearch.referencedIndexes,
-                                               queryContext);
+            return TermIterator.build(sstableIntersections, memtableResults, queryView.referencedIndexes, queryContext);
         }
         catch (Throwable t)
         {
             // all sstable indexes in view have been referenced, need to clean up when exception is thrown
             queryView.referencedIndexes.forEach(SSTableIndex::release);
-            // if ANN sstable indexes are referenced separately, release them
-            if (annQueryViewInHybridSearch != null)
-                annQueryViewInHybridSearch.referencedIndexes.forEach(SSTableIndex::release);
-
             throw t;
         }
     }
 
-    private RangeIterator<PrimaryKey> convertToPrimaryKeyIterator(PrimaryKeyMap.Factory pkFactory, RangeIterator<Long> sstableRowIdsIterator)
+    // This is a hybrid query. We apply all other predicates before ordering and limiting.
+    public RangeIterator getTopKRows(RangeIterator source, RowFilter.Expression expression)
     {
-        if (sstableRowIdsIterator.getCount() <= 0)
-            return RangeIterator.emptyKeys();
-
-        PrimaryKeyMap primaryKeyMap = pkFactory.newPerSSTablePrimaryKeyMap();
-        return SSTableRowIdKeyRangeIterator.create(primaryKeyMap, queryContext, sstableRowIdsIterator);
+        return new OrderingFilterRangeIterator(source,
+                                               ORDER_CHUNK_SIZE,
+                                               list -> this.getTopKRows(list, expression));
     }
 
-    private RangeIterator<PrimaryKey> reorderAndLimitBy(RangeIterator<PrimaryKey> original, Memtable memtable, Expression expression)
+    private RangeIterator getTopKRows(List<PrimaryKey> rawSourceKeys, RowFilter.Expression expression)
     {
-        return expression.context.reorderMemtable(memtable, queryContext, original, expression, getLimit());
+        // Filter out PKs now. Each PK is passed to every segment of the ANN index, so filtering shadowed keys
+        // eagerly can save some work when going from PK to row id for on disk segments.
+        // Since the result is shared with multiple streams, we use an unmodifiable list.
+        var sourceKeys = rawSourceKeys.stream().filter(queryContext::shouldInclude).collect(Collectors.toUnmodifiableList());
```

Comment:
toUmodifiableList is fairly heavyweight (essentially `List.of(List.toArray())`).  not sure if it's a big deal here or not.

Reply:
Let's remove it then. This was only defensive, but we control the stack here, so it's not worth the cost.